arangod/Replication2/ReplicatedState/ReplicatedState.tpp

```diff
@@ -252,10 +251,38 @@ auto ReplicatedStateManager<S>::resign() && -> std::unique_ptr<CoreType> {
   return std::move(core);
 }
 
-template<typename EntryType, typename Deserializer,
-         template<typename> typename Interface, typename ILogMethodsT>
-auto StreamProxy<EntryType, Deserializer, Interface,
-                 ILogMethodsT>::waitForIterator(LogIndex index)
+template<typename S, template<typename> typename Interface,
+         ValidStreamLogMethods ILogMethodsT>
+StreamProxy<S, Interface, ILogMethodsT>::StreamProxy(
+    std::unique_ptr<ILogMethodsT> methods)
+    : _logMethods(std::move(methods)) {}
+
+template<typename S, template<typename> typename Interface,
+         ValidStreamLogMethods ILogMethodsT>
+auto StreamProxy<S, Interface, ILogMethodsT>::methods() -> auto& {
+  return *this->_logMethods;
+}
+
+template<typename S, template<typename> typename Interface,
+         ValidStreamLogMethods ILogMethodsT>
+auto StreamProxy<S, Interface,
+                 ILogMethodsT>::resign() && -> decltype(_logMethods) {
+  return std::move(this->_logMethods);
+}
+
+template<typename S, template<typename> typename Interface,
+         ValidStreamLogMethods ILogMethodsT>
+auto StreamProxy<S, Interface, ILogMethodsT>::waitFor(LogIndex index)
+    -> futures::Future<WaitForResult> {
+  // TODO As far as I can tell right now, we can get rid of this:
+  //      Delete this, also in streams::Stream.
+  return _logMethods->waitFor(index).thenValue(
+      [](auto const&) { return WaitForResult(); });
```

Comment:
`waitFor` is used in `DocumentLeaderState::replicateOperation`, in case we have to wait for commit (line 181).

Reply:
> Apart from `waitFor`, we also use `stream->insert()` in `DocumentLeaderState::replicateOperation`. Should we also throw an exception for them as well?

I'm unsure. It's not necessary, as `insert` can not be called after the state resigned, and the stream is purposefully resigned only after the state. So it'd also be plausible to add an assertion that `insert` is not called after being resigned. On the other hand, maybe it's better to have a consistent interface on the stream regarding resign. What's your take on this?